File: go/day01/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"sort"
	"strconv"
	"strings"
)

var input string

func init() {
	content, err := ioutil.ReadFile("input.txt")
	if err != nil {
		log.Fatal(err)
	}

	input = strings.TrimRight(string(content), "\n")
}

func main() {
	ans1 := part1(input)
	fmt.Println("Part 1:", ans1)

	ans2 := part2(input)
	fmt.Println("Part 2:", ans2)
}

func part1(input string) int {
	max := 0

	for _, entity := range strings.Split(input, "\n\n") {
		current := 0
		for _, calorie := range strings.Split(entity, "\n") {
			calorie, err := strconv.Atoi(calorie)
			if err != nil {
				log.Fatal(err)
			}
			current += calorie
		}

		if current > max {
			max = current
		}
	}

	return max
}

func part2(input string) int {
	max_arr := []int{}

	for _, entity := range strings.Split(input, "\n\n") {
		current := 0
		for _, calorie := range strings.Split(entity, "\n") {
			calorie, err := strconv.Atoi(calorie)
			if err != nil {
				log.Fatal(err)
			}
			current += calorie
		}
		max_arr = append(max_arr, current)
	}

	sort.Ints(max_arr)
	sum := 0
	for _, num := range max_arr[len(max_arr)-3:] {
		sum += num
	}

	return sum
}

File: zig/day01/main.zig
const std = @import("std");

pub fn main() !void {
    const file = @embedFile("input.txt");
    const part1_result = try solvePart1(file);
    std.debug.print("Part 1: {}\n", .{part1_result});

    const part2_result = try solvePart2(file);
    std.debug.print("Part 2: {}\n", .{part2_result});
}

pub fn solvePart1(input: []const u8) !u32 {
    var lines = std.mem.split(u8, input, "\n");
    var max: u32 = 0;
    var current: u32 = 0;

    while (lines.next()) |line| {
        if (line.len == 0) {
            if (current > max) {
                max = current;
            }
            current = 0;
        } else {
            const value = try std.fmt.parseInt(u32, line, 10);
            current += value;
        }
    }
    return max;
}

pub fn solvePart2(input: []const u8) !u32 {
    var lines = std.mem.split(u8, input, "\n");
    var max = [3]u32{ 0, 0, 0 };
    var current: u32 = 0;

    while (lines.next()) |line| {
        if (line.len == 0) {
            for (max, 0..) |n, i| {
                if (current > n) {
                    if (i == 0) {
                        max[2] = max[1];
                        max[1] = max[0];
                        max[0] = current;
                    } else if (i == 1) {
                        max[2] = max[1];
                        max[1] = current;
                    } else {
                        max[2] = current;
                    }
                    break;
                }
            }
            current = 0;
        } else {
            const value = try std.fmt.parseInt(u32, line, 10);
            current += value;
        }
    }
    return @reduce(.Add, @as(@Vector(3, u32), max));
}

test "test-part1" {
    const file = @embedFile("test.txt");
    const result_part_1 = try solvePart1(file);
    try std.testing.expect(result_part_1 == 24000);
}

test "test-part2" {
    const file = @embedFile("test.txt");
    const result_part_2 = try solvePart2(file);
    try std.testing.expect(result_part_2 == 45000);
}
